test/e2e/storage/pv_protection.go

```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/slice"
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+var _ = utils.SIGDescribe("PV Protection [Feature:StorageProtection]", func() {
+	var (
+		client    clientset.Interface
+		nameSpace string
+		err       error
+		pvc       *v1.PersistentVolumeClaim
+		pv        *v1.PersistentVolume
+		pvConfig  framework.PersistentVolumeConfig
+		pvcConfig framework.PersistentVolumeClaimConfig
+		volLabel  labels.Set
+		selector  *metav1.LabelSelector
+	)
+
+	f := framework.NewDefaultFramework("pv-protection")
+	BeforeEach(func() {
+		client = f.ClientSet
+		nameSpace = f.Namespace.Name
+		framework.ExpectNoError(framework.WaitForAllNodesSchedulable(client, framework.TestContext.NodeSchedulableTimeout))
+
+		// Enforce binding only within test space via selector labels
+		volLabel = labels.Set{framework.VolumeSelectorKey: nameSpace}
+		selector = metav1.SetAsLabelSelector(volLabel)
+
+		pvConfig = framework.PersistentVolumeConfig{
+			NamePrefix: "hostpath-",
+			Labels:     volLabel,
+			PVSource: v1.PersistentVolumeSource{
+				HostPath: &v1.HostPathVolumeSource{
+					Path: "/tmp/data",
+				},
+			},
+		}
+
+		pvcConfig = framework.PersistentVolumeClaimConfig{
+			Annotations: map[string]string{
+				v1.BetaStorageClassAnnotation: "",
+			},
+			Selector: selector,
+		}
+
+		By("Creating a PV")
+		// make the pv definitions
+		pv = framework.MakePersistentVolume(pvConfig)
+		// create the PV
+		pv, err := client.CoreV1().PersistentVolumes().Create(pv)
```

Comment:
`:=` creates a new local `pv` variable, so the tests and `AfterEach` see `pv` as nil and everything fails. Did you run the test? I did ;-)

Reply:
sorry for that. i did not run this locally, will do it

Reply:
The `pv` variable is initialized by `pv = framework.MakePersistentVolume(pvConfig)`, in the above line, so I'm not sure if `AfterEach` fails or not. But we really should use `=` instead of `:=`.